service/src/main/java/bio/terra/workspace/service/job/JobService.java

```diff
@@ -394,15 +397,30 @@ private FlightMap getResultMap(FlightState flightState) {
     return resultMap;
   }
 
+  /**
+   * Ensure the user in the user request has permission to read the workspace associated with the
+   * Job ID. Throws ForbiddenException if not. If the workspace does not exist, treat this as the
+   * user having access (i.e. skip validateWorkspaceAndAction check).
+   *
+   * @param jobId - ID of running job
+   * @param userRequest - original user request
+   */
   private void verifyUserAccess(String jobId, AuthenticatedUserRequest userRequest) {
     try {
       FlightState flightState = stairwayComponent.get().getFlightState(jobId);
       FlightMap inputParameters = flightState.getInputParameters();
-      String flightSubjectId =
-          inputParameters.get(JobMapKeys.SUBJECT_ID.getKeyName(), String.class);
-      if (!StringUtils.equals(flightSubjectId, userRequest.getSubjectId())) {
-        throw new JobUnauthorizedException("Unauthorized");
-      }
+      UUID workspaceId = inputParameters.get(WorkspaceFlightMapKeys.WORKSPACE_ID, UUID.class);
+
+      // If the workspace doesn't exist, don't try to validate it
```

Comment:
I'm not totally following why we need this exception. Is the problem that we currently have tests which delete a workspace and then read jobs from that workspace afterwards, which are now failing? 

403 seems like a fine behavior for queries reading jobs from a deleted workspace. If that's the real behavior we want, could we update tests to reflect that instead?

Reply:
The semantics of all of our deletes is to return success if we find the object already deleted. The thought behind that is to make it easier for callers to simply retry rather than interpret the error.

Reply:
The main problem I saw was that deleteWorkspace went from being a no-op when the workspace wasn't found to throwing `MissingWorkspaceException`. This broke all kinds of tests.

Reply:
I don't think that's currently our semantic for deleting workspaces, right now we throw a 404 for workspaces which are already deleted.

I think the problem is that without this check `WorkspaceService. deleteWorkspace` will always fail, not just when the workspace wasn't found (which is current behavior). The problem is that `deleteWorkspace` can't read the result of the job it submits because the workspace is gone.

Instead of allowing anyone to read jobs from deleted workspaces, I think it would be cleaner to fix `WorkspaceService. deleteWorkspace`. Either:

- This method calls a new JobService endpoint which gets the job result without doing an auth check. `deleteWorkspace` already verifies the user's permission before doing the delete
- This method catches a `WorkspaceNotFoundException` from the `submitAndWait` call and treats it as success
- Something else?